api/src/convert/tx.rs

```diff
@@ -51,69 +50,144 @@ mod tests {
+            {
+                let bytes = mc_util_serial::encode(&tx);
+                let recovered_tx = external::Tx::parse_from_bytes(&bytes).unwrap();
+                assert_eq!(recovered_tx, external::Tx::from(&tx));
+            }
+
+            // Encoding with protobuf, decoding with prost should be the identity function.
+            {
+                let external_tx: external::Tx = external::Tx::from(&tx);
+                let bytes = external_tx.write_to_bytes().unwrap();
+                let recovered_tx: Tx = mc_util_serial::decode(&bytes).unwrap();
+                assert_eq!(tx, recovered_tx);
+            }
+        }
+    }
+
+    #[test]
+    /// Tx --> externalTx --> Tx should be the identity function, for tx with
+    /// input rules
+    fn test_convert_tx_with_input_rules() {
+        // Generate a Tx to test with. This is copied from
+        // transaction_builder.rs::test_simple_transaction
+        let mut rng: StdRng = SeedableRng::from_seed([1u8; 32]);
+
+        for block_version in BlockVersion::iterator().skip(3) {
+            let alice = AccountKey::random(&mut rng);
+            let bob = AccountKey::random(&mut rng);
+            let charlie = AccountKey::random(&mut rng);
+
+            let token2 = TokenId::from(2);
+
+            let fpr = MockFogResolver::default();
+
+            // Charlie makes a signed contingent input, offering 1000 token2's for 1 MOB
+            let input_credentials = get_input_credentials(
+                block_version,
+                Amount::new(1000, token2),
+                &charlie,
+                &fpr,
+                &mut rng,
+            );
+            let mut sci_builder = SignedContingentInputBuilder::new(
+                block_version,
+                input_credentials,
+                vec![59],
+                fpr.clone(),
+                EmptyMemoBuilder::default(),
             );
 
-            let membership_proofs: Vec<TxOutMembershipProof> = ring
-                .iter()
-                .map(|_tx_out| {
-                    // TransactionBuilder does not validate membership proofs, but does require one
-                    // for each ring member.
-                    TxOutMembershipProof::new(0, 0, Default::default())
-                })
-                .collect();
-
-            let input_credentials = InputCredentials::new(
-                ring.clone(),
-                membership_proofs,
-                0,
-                onetime_private_key,
-                *alice.view_private_key(),
+            sci_builder
+                .add_output(
+                    Amount::new(1000 * MILLIMOB_TO_PICOMOB, Mob::ID),
+                    &charlie.default_subaddress(),
+                    &mut rng,
+                )
+                .unwrap();
+
+            let sci = sci_builder.build(&mut rng).unwrap();
+
+            // Alice sends this token2 amount to Bob from Charlie, paying Charlie 1000 MOB
```

Comment:
Nit: Alice pays Charlie only 1.475 Mob and getting change for .475 Mob. Alice is paying Bob 1000 Token2.